[NumberField] `snapOnStep` prop

atomiks opened the pull request:
Closes #1542
Closes #1010

The new `snapOnStep` prop is a boolean:

- `true`: the value is snapped to the nearest step when stepping (incrementing/decrementing). So `1.07` with a step of `1` becomes `2` when incrementing
- `false` (new default): the value isn't snapped at all. So `1.07` with a step of `1` becomes `2.07`.

Also new behavior for `true`:

- If you type a decimal then increment/decrement, it snaps to the nearest regular `step`, just like `largeStep` already did.
- It will snap the nearest step rather than round and unexpectedly go "more" than wanted.

michaldudak commented:
Using small or large step with the custom function doesn't work. Decrement doesn't work at all, while increment adds a double value. See https://codesandbox.io/p/devbox/silly-moser-t85lpm?file=%2Fsrc%2FApp.tsx%3A12%2C22&workspaceId=ws_QAeLcZEcDnfGaFA47etEUH

atomiks commented:
@michaldudak sandbox is private

michaldudak commented:
Try now

michaldudak commented:
It now works fine with `stepBehavior="snap"` for small and large steps, but the normal step doesn't snap (10.1 + step = 11.1)

atomiks commented:
@michaldudak yeah - hadn't yet changed the implementation after our chat. Just changed the math in a new commit and updated PR description with additional details
